libsolidity/analysis/FunctionCallGraph.cpp

```diff
@@ -0,0 +1,219 @@
+/*
+	This file is part of solidity.
+
+	solidity is free software: you can redistribute it and/or modify
+	it under the terms of the GNU General Public License as published by
+	the Free Software Foundation, either version 3 of the License, or
+	(at your option) any later version.
+
+	solidity is distributed in the hope that it will be useful,
+	but WITHOUT ANY WARRANTY; without even the implied warranty of
+	MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+	GNU General Public License for more details.
+
+	You should have received a copy of the GNU General Public License
+	along with solidity.  If not, see <http://www.gnu.org/licenses/>.
+*/
+// SPDX-License-Identifier: GPL-3.0
+
+#include <libsolidity/analysis/FunctionCallGraph.h>
+#include <boost/range/adaptor/reversed.hpp>
+
+using namespace std;
+using namespace solidity::frontend;
+
+bool FunctionCallGraphBuilder::CompareByID::operator()(Node const& _lhs, Node const& _rhs) const
+{
+	if (_lhs.index() != _rhs.index())
+		return _lhs.index() < _rhs.index();
+
+	if (std::holds_alternative<SpecialNode>(_lhs))
+		return std::get<SpecialNode>(_lhs) < std::get<SpecialNode>(_rhs);
+	return std::get<ASTNode const*>(_lhs)->id() < std::get<ASTNode const*>(_rhs)->id();
+}
+
+bool FunctionCallGraphBuilder::CompareByID::operator()(Node const& _lhs, int64_t _rhs) const
+{
+	solAssert(!std::holds_alternative<SpecialNode>(_lhs), "");
+
+	return std::get<ASTNode const*>(_lhs)->id() < _rhs;
+}
+
+bool FunctionCallGraphBuilder::CompareByID::operator()(int64_t _lhs, Node const& _rhs) const
+{
+	solAssert(!std::holds_alternative<SpecialNode>(_rhs), "");
+
+	return _lhs < std::get<ASTNode const*>(_rhs)->id();
+}
+
+unique_ptr<FunctionCallGraphBuilder::ContractCallGraph> FunctionCallGraphBuilder::create(ContractDefinition const& _contract)
+{
+	FunctionCallGraphBuilder builder;
+
+	builder.m_contract = &_contract;
+
+	builder.m_graph = make_unique<ContractCallGraph>(_contract);
```

Comment:
So I take it you *don't* want to move all of it to the constructor :-)?

Reply:
Well, not all as I mentioned, but I think you need to look at https://github.com/ethereum/solidity/pull/10332/commits/28e72867b35b4a0dc2b01e614165a12199d0daf4#diff-4d1d9decb6ec37cd1826af7c0004fc0a94b3e375c59b05a07a0ace0323f67060R51

Reply:
You didn't want to move the initialization of the members to in-class member-initialization from what I understood - but I think we may have been talking past each other a bit :-). The idea would have been to just have
```
unique_ptr<FunctionCallGraphBuilder::ContractCallGraph> FunctionCallGraphBuilder::create(ContractDefinition const& _contract)
{
	FunctionCallGraphBuilder builder(_contract);
        return builder.m_graph;
}
```
and put everything that's currently in here in the body of the constructor of ``FunctionCallGraphBuilder`` - the advantage being that you can get rid of all the ``builder.`` qualifiers throughout this function, which I'd say would be nicer.
But I'd also be fine with not doing that, so as you like :-). I'll take a look at the rest of the PR.

Reply:
I think it was all just a big misunderstanding :D
Your idea sounds fine to me. Will change.

Reply:
Yeah, sorry, I think I should have been clearer in my initial suggestion :-).

Reply:
@ekpyron I thought that having very little work in the constructor (only initializing members basically) and keeping most of the work in the factory method was a nicer design?

I'm not sure how much the factory method was doing here before (i.e. just a few lines more of all of the stuff that's in the constructor now?) but I was thinking about moving all the visiting to this method. That would also work better with the new architecture - having two separate factory methods, one for creating the creation graph and other for deployment graph would work better than having two separate constructors or one constructor with a flag.

What do you think?

Reply:
There's little need to consider these old comments (even more so since it's quite hard to find, display and see them on github - or respond to them, I actually have a delay while typing right now :-)). Since these comments the overall design changed significantly so none of it will really apply anymore.

Reply:
Most of the old ones are already resolved but some that aren't do seem still relevant.

As for the slowness - it's fine on my side after the page is fully loaded, just tedious to unfold stuff. But if it's slow for you, maybe I should just close this PR and open a new one?